fix grad silent correctness issue from view fn followed by an inplace fn

samdow opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #85374

From https://github.com/pytorch/functorch/issues/1007, which was an issue where we would wrap aliases of unwrapped tensors and miss the inplace error message where we should have gotten it. Instead of keeping aliases unwrapped like I had originally wanted, this simplifies it slightly such that:
(1) All tensors that were previously wrapped are still wrapped. This is occasionally important because of the 1-1 relationship between a tensor and autograd meta. By keeping the same number of wrapper tensors before, we'll never have autograd try to write multiple autograd metas to the same tensor when it wouldn't before
(2) The tensors that either were unwrapped tensors or aliases of unwrapped tensors now get a flag on them (now called `alias_of_unwrapped`). This way, they are still wrapper tensors (and don't have to potentially break autograd) but we can identify that they should be treated like an unwrapped tensor

Commit:
fix grad silent correctness issue from view fn followed by an inplace fn

[ghstack-poisoned]

Commit:
Update on "fix grad silent correctness issue from view fn followed by an inplace fn"


From https://github.com/pytorch/functorch/issues/1007

This adds the restriction so that if an input is (1) unwrapped at the current level and (2) is an alias of a returned value, we won't wrap that returned value

However, even unwrapped tensors need to be wrapped by the DyanmicLayerFrontMode key. So, this adds a "generated" value to the TensorWrapper that is true only if the tensor was created by the DyanmicLayerFrontMode because the original tensor was unwrapped. Then, we skip wrapping all the outputs that are aliased to inputs that were either unwrapped or generated

[ghstack-poisoned]

zou3519 commented:
Test failures look legit

zou3519 commented on functorch/csrc/TensorWrapper.h:
```diff
@@ -38,6 +38,7 @@ struct FUNCTORCH_API TensorWrapper : public c10::TensorImpl {
       Tensor value,
       int64_t level,
       std::shared_ptr<bool> is_alive,
+      bool generated = false,  // if false underlying was passed as input to this level, otherwise wrapper was generated by interpreter
```
> However, even unwrapped tensors need to be wrapped by the DyanmicLayerFrontMode key. So, this adds a "generated" value to the TensorWrapper that is true only if the tensor was created by the DyanmicLayerFrontMode because the original tensor was unwrapped. Then, we skip wrapping all the outputs that are aliased to inputs that were either unwrapped or generated

I don't understand why the `generated` is necessary: why do unwrapped tensors need to be wrapped by the DynamicLayerFrontMode key?

zou3519 replied:
Is it necessary for the case where some inputs are wrapped and some inputs aren't wrapped?

samdow replied:
Answering in reverse order
> why do unwrapped tensors need to be wrapped by the DynamicLayerFrontMode key? Is it necessary for the case where some inputs are wrapped and some inputs aren't wrapped?

Yes it should only come up where some inputs are wrapped and others aren't. The unwrapped ones are wrapped right before they run through autograd so that the AutogradMeta information gets to go on the right tensor. Also if the unwrapped tensor has any autograd metadata already on it, autograd gets mad if it sees the unwrapped version since autograd meta and tensors are 1-1

> I don't understand why the generated is necessary

This might have already been clear but just for extra clarity, once the tensors are wrapped, they all look the same so I was thinking that we have to pass the information about which tensors were originally wrapped from the transform to the send to next interpreter functions. Since the interpreters don't currently have state that's not like the constant state of the transform, it felt easier to attach it to the tensors themselves

Commit:
Update on "fix grad silent correctness issue from view fn followed by an inplace fn"


From https://github.com/pytorch/functorch/issues/1007

This adds the restriction so that if an input is (1) unwrapped at the current level and (2) is an alias of a returned value, we won't wrap that returned value

However, even unwrapped tensors need to be wrapped by the DyanmicLayerFrontMode key. So, this adds a "generated" value to the TensorWrapper that is true only if the tensor was created by the DyanmicLayerFrontMode because the original tensor was unwrapped. Then, we skip wrapping all the outputs that are aliased to inputs that were either unwrapped or generated

[ghstack-poisoned]

Commit:
Update on "fix grad silent correctness issue from view fn followed by an inplace fn"


From https://github.com/pytorch/functorch/issues/1007

This adds the restriction so that if an input is (1) unwrapped at the current level and (2) is an alias of a returned value, we won't wrap that returned value

However, even unwrapped tensors need to be wrapped by the DyanmicLayerFrontMode key. So, this adds a "generated" value to the TensorWrapper that is true only if the tensor was created by the DyanmicLayerFrontMode because the original tensor was unwrapped. Then, we skip wrapping all the outputs that are aliased to inputs that were either unwrapped or generated

[ghstack-poisoned]

Commit:
Update on "fix grad silent correctness issue from view fn followed by an inplace fn"


From https://github.com/pytorch/functorch/issues/1007, which was an issue where we would wrap aliases of unwrapped tensors and miss the inplace error message where we should have gotten it. Instead of keeping aliases unwrapped like I had originally wanted, this simplifies it slightly such that:
(1) All tensors that were previously wrapped are still wrapped. This is occasionally important because of the 1-1 relationship between a tensor and autograd meta. By keeping the same number of wrapper tensors before, we'll never have autograd try to write multiple autograd metas to the same tensor when it wouldn't before
(2) The tensors that either were unwrapped tensors or aliases of unwrapped tensors now get a flag on them (now called `alias_of_unwrapped`). This way, they are still wrapper tensors (and don't have to potentially break autograd) but we can identify that they should be treated like an unwrapped tensor

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/functorch/TensorWrapper.h:
```diff
@@ -89,7 +89,7 @@ struct TORCH_API TensorWrapper : public c10::TensorImpl {
 };
 
-TORCH_API Tensor makeTensorWrapper(const Tensor& tensor, int64_t level, bool generated=false);
+TORCH_API Tensor makeTensorWrapper(const Tensor& tensor, int64_t level, bool alias_of_unwrapped=false);
```
I'm still thinking through this PR. Here's a brain dump of where I'm at:
- The wrap-call-unwrap we are doing in this PR. I see why it is necessary, but it makes me nervous. One way around it is that we say if we are doing a view operation and ALL tensors are not wrapped, then we ignore the current grad layer. This is what we do for vmap (if the tensors are not batched at the current level, then ignore the current vmap layer). If we are doing a view operation with some wrapped tensors and some unwrapped tensors, then we can error out.
- There's a BC-breaking vs silently incorrect behavior tradeoff. I don't want to accidentally make some behavior that people do rely on today error out.
- Performance. Checking alias annotations on schemas on every dispatch call might be a problem
- in-place view operations: do we have a problem with those?

zou3519 replied:
Another way around this is: we introduce some `is_immutable` flag to TensorWrapper. Tensors that get automatically wrapped in a TensorWrapper get the `is_immutable` flag. View operations propagate the `is_immutable`; out-of-place operations don't propagate it, and in-place operations cause an error if the TensorWrapper is immutable.

Pros: No more BC-breaking vs silently incorrect behavior tradeoff. Only silently incorrect behavior will be disallowed
Cons: Could be difficult to implement

samdow replied:
> Another way around this is: we introduce some is_immutable flag to TensorWrapper and keep the existing wrapping logic. Tensors that get automatically wrapped in a TensorWrapper get the is_immutable flag. View operations propagate the is_immutable; out-of-place operations don't propagate it, and in-place operations cause an error if the TensorWrapper is immutable.

I'm missing--how is that different than what's done here?

The previous name (generated) was a poor choice, so in case this was unclear: Originally I was thinking that we could just leave aliased tensors unwrapped. However, this caused some issues with the autograd meta. So, I ended up leaving the wrapping logic exactly as it was but if a tensor was unwrapped when it got to the front layer, the wrapper generated for it would have a flag on it (currently called `alias_of_unwrapped`) and that flag would only be propagated if the output is an alias of the input (based on the schema's alias info). If we tried to do run an inplace op on a tensor with this flag, we fail

> Performance. Checking alias annotations on schemas on every dispatch call might be a problem

Yeah this is a problem. I was thinking that we could memoize the function to start with? In the future, we might also be able to add this as a property that we compute when we build the schemas but I don't know the function schema code super well

> in-place view operations: do we have a problem with those?

To check, you're talking about ops marked with the `inplace_view` tag. For those, they match the [inplace op logic](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/functorch/DynamicLayer.cpp#L353) so they will have [always failed with an unwrapped input](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/functorch/ADInterpreters.cpp#L7)

zou3519 replied:
> I'm missing--how is that different than what's done here?

You're right -- that's exactly what this PR is doing! I was confused because I didn't read the PR, I was going off of what I thought the design was, which was that we were attempting to return an unwrapped Tensor if the aliased tensor was unwrapped

zou3519 replied:
> > in-place view operations: do we have a problem with those?
> 
> To check, you're talking about ops marked with the inplace_view tag. For those, they match the [inplace op logic](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/functorch/DynamicLayer.cpp#L353) so they will have [always failed with an unwrapped input](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/functorch/ADInterpreters.cpp#L7)

Yes, that's a good point.

So the problem right now is figuring out the performance

zou3519 commented on aten/src/ATen/functorch/ADInterpreters.cpp:
```diff
@@ -26,19 +26,22 @@ static void checkForInvalidMutationOnCaptures(
 }
 
 static Tensor materializeGradWrappers(const Tensor& tensor, int64_t current_level) {
+  // we will only mark something as an alias of an unwrapped tensor if it's during user code in a transform.
+  // This state is saved by the during functorch transform context manager
+  const bool mark_as_alias_of_unwrapped = getDuringFunctorchTransform();
```
Why is the context manager necessary? Is it a performance optimization or is it for correctness or something else?

samdow replied:
It's for correctness--during the setup for the transforms (particularly I think like building the dual in jvp), we end up calling some operations that alias the unwrapped input, but it's part of the code to build the wrapper. I'll clarify the comment

zou3519 replied:
Off of the top of your head, do you remember which tests failed without this? I'm trying to get a repro to dig in more

samdow replied:
Somehow I thought it was `vmapvjpvjp` or `jvpvjp` but now I'm not seeing it locally--I have some small naming/comment local changes anyway so I'll just removed this context manager and change and we can see if anything breaks in CI

samdow replied:
but basically it might have gotten caught up with a couple of other issues, sorry if you've already started investigating!

zou3519 replied:
Nah, I just ran the tests locally and see if anything would fail. The tests finished running and everything seems to pass

zou3519 commented on aten/src/ATen/functorch/ADInterpreters.cpp:
```diff
@@ -136,12 +156,29 @@ static void autogradBasedTransformSendToNext(
+  std::vector<int64_t> unwrapped_outputs; // indexes of outputs that should remain unwrapped
+
+  // lift_fresh: it's must be freshly allocated and should be wrapped. User shouldn't have access to input version
+  // alias: this is needed for the CompositeImplicit instance norm (running_mean/var get set to be a wrapped value)
+  //        It's not a user facing function, but is more prone to possible errors
```
Why is there a special case for alias?

zou3519 replied:
Do you remember what tests failed without this? (I want to step into it a little)

samdow replied:
This one is actually saved on hud (bless ghstack)

Tests: 
* `test_vjpvjp_nn_functional_instance_norm`
* `test_jvpvjp_nn_functional_instance_norm`

Error:
`RuntimeError: During a grad (vjp, jvp, grad, etc) transform, the function provided attempted to call in-place operation (aten::copy_) that would mutate a captured Tensor. This is not supported; please rewrite the function being transformed to explicitly accept the mutated Tensor(s) as inputs.`
<details>
<summary>Full stacktrace</summary>

```
Traceback (most recent call last):
2022-09-22T00:19:00.5652985Z   File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 391, in instantiated_test
2022-09-22T00:19:00.5653466Z     raise rte
2022-09-22T00:19:00.5654132Z   File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 378, in instantiated_test
2022-09-22T00:19:00.5691739Z     result = test(self, **param_kwargs)
2022-09-22T00:19:00.5692673Z   File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/common_device_type.py", line 815, in test_wrapper
2022-09-22T00:19:00.5693207Z     return test(*args, **kwargs)
2022-09-22T00:19:00.5693656Z   File "/var/lib/jenkins/workspace/functorch/test/test_ops.py", line 1218, in test_jvpvjp
2022-09-22T00:19:00.5694195Z     result = jvp(push_vjp, (primals, cotangents), (primals_tangents, cotangents_tangents))
2022-09-22T00:19:00.5694912Z   File "/opt/conda/lib/python3.7/site-packages/functorch/_src/eager_transforms.py", line 801, in jvp
2022-09-22T00:19:00.5695519Z     return _jvp_with_argnums(func, primals, tangents, argnums=None, strict=strict, has_aux=has_aux)
2022-09-22T00:19:00.5696288Z   File "/opt/conda/lib/python3.7/site-packages/functorch/_src/eager_transforms.py", line 850, in _jvp_with_argnums
2022-09-22T00:19:00.5696741Z     result_duals = func(*duals)
2022-09-22T00:19:00.5697174Z   File "/var/lib/jenkins/workspace/functorch/test/test_ops.py", line 1215, in push_vjp
2022-09-22T00:19:00.5697592Z     _, vjp_fn = vjp(fn, *primals)
2022-09-22T00:19:00.5698184Z   File "/opt/conda/lib/python3.7/site-packages/functorch/_src/eager_transforms.py", line 273, in vjp
2022-09-22T00:19:00.5698699Z     return _vjp_with_argnums(func, *primals, has_aux=has_aux)
2022-09-22T00:19:00.5699402Z   File "/opt/conda/lib/python3.7/site-packages/functorch/_src/eager_transforms.py", line 300, in _vjp_with_argnums
2022-09-22T00:19:00.5699984Z     primals_out = func(*primals)
2022-09-22T00:19:00.5700405Z   File "/var/lib/jenkins/workspace/functorch/test/test_ops.py", line 118, in wrapped
2022-09-22T00:19:00.5701065Z     result = f(*_args, **kwargs)
2022-09-22T00:19:00.5701762Z   File "/opt/conda/lib/python3.7/site-packages/torch/testing/_internal/opinfo/core.py", line 1026, in __call__
2022-09-22T00:19:00.5702217Z     return self.op(*args, **kwargs)
2022-09-22T00:19:00.5702830Z   File "/opt/conda/lib/python3.7/site-packages/torch/nn/functional.py", line 2492, in instance_norm
2022-09-22T00:19:00.5703459Z     input, weight, bias, running_mean, running_var, use_input_stats, momentum, eps, torch.backends.cudnn.enabled
2022-09-22T00:19:00.5704931Z RuntimeError: During a grad (vjp, jvp, grad, etc) transform, the function provided attempted to call in-place operation (aten::copy_) that would mutate a captured Tensor. This is not supported; please rewrite the function being transformed to explicitly accept the mutated Tensor(s) as inputs.
```

</details>

https://hud.pytorch.org/pytorch/pytorch/pull/85374?sha=9c002fb612fa90a879280181c59df16fee9f81ca

samdow replied:
Also I'm pretty sure the copy_ is coming from here
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Normalization.cpp#L633

zou3519 commented on aten/src/ATen/functorch/ADInterpreters.cpp:
```diff
@@ -136,12 +156,29 @@ static void autogradBasedTransformSendToNext(
+  // lift_fresh: it's must be freshly allocated and should be wrapped. User shouldn't have access to input version
+  // alias: this is needed for the CompositeImplicit instance norm (running_mean/var get set to be a wrapped value)
+  //        It's not a user facing function, but is more prone to possible errors
+  const bool always_wrap = op.schema().name() == "aten::lift_fresh" || op.schema().name() == "aten::alias";
```
String comparison is slow

samdow replied:
Locally I'm seeing that with 10M tensor creations it goes from ~59s with this check to ~54s without (replacing this with a constant true). I upped it to try to get rid of too much noise but basically having this is ~10% slowdown per call

zou3519 replied:
We can do something like what we previously did with the JVP decomposition. There, we also needed to do string comparison, but I was worried it would be too slow, so we added a workaround. The workaround was something like:
- ATenOpId is an enum with various operators
- DynamicLayerFront gets a registration for each operator that we care about (in this case, alias and lift_fresh). These kernels will pass the correct ATenOpId down the line
- We can give each Interpreter::process a Interpreter::process_with_id that takes an additional ATenOpId and does something with it


Code for reference:
https://github.com/pytorch/pytorch/commit/f10e9d66c49c023bdf1a819a10b327daaeb161c5#diff-edf34e92e4e457dea93693fc196c7a6af5ec14d7e9a2f9aa51529d6d7427517aR369

Long-term it sounds like we need some sort of string interning for dispatcher operations. Let me write up an issue on it. It might be doable to write that in a couple of days and that would be better than if we cooked up the hack above

samdow replied:
Updated! I made it more simple and specialized (sendToNext now just accepts an extra bool) than the forward over reverse example since both lift_fresh and alias have the same behavior. There's a note there though that if we need more special behavior than just this, it's going to scale better to use enums

zou3519 replied:
Approach SGTM, will read the code closer soon. Do you also have microbenchmarks for this new approach?

samdow replied:
Small enough to probably just be noise (for 10M calls)
Without this check: 54.707358837127686s
With: 54.98072600364685s

zou3519 replied:
That's awesome

Commit:
Update on "fix grad silent correctness issue from view fn followed by an inplace fn"


From https://github.com/pytorch/functorch/issues/1007, which was an issue where we would wrap aliases of unwrapped tensors and miss the inplace error message where we should have gotten it. Instead of keeping aliases unwrapped like I had originally wanted, this simplifies it slightly such that:
(1) All tensors that were previously wrapped are still wrapped. This is occasionally important because of the 1-1 relationship between a tensor and autograd meta. By keeping the same number of wrapper tensors before, we'll never have autograd try to write multiple autograd metas to the same tensor when it wouldn't before
(2) The tensors that either were unwrapped tensors or aliases of unwrapped tensors now get a flag on them (now called `alias_of_unwrapped`). This way, they are still wrapper tensors (and don't have to potentially break autograd) but we can identify that they should be treated like an unwrapped tensor

[ghstack-poisoned]

Commit:
Update on "fix grad silent correctness issue from view fn followed by an inplace fn"


From https://github.com/pytorch/functorch/issues/1007, which was an issue where we would wrap aliases of unwrapped tensors and miss the inplace error message where we should have gotten it. Instead of keeping aliases unwrapped like I had originally wanted, this simplifies it slightly such that:
(1) All tensors that were previously wrapped are still wrapped. This is occasionally important because of the 1-1 relationship between a tensor and autograd meta. By keeping the same number of wrapper tensors before, we'll never have autograd try to write multiple autograd metas to the same tensor when it wouldn't before
(2) The tensors that either were unwrapped tensors or aliases of unwrapped tensors now get a flag on them (now called `alias_of_unwrapped`). This way, they are still wrapper tensors (and don't have to potentially break autograd) but we can identify that they should be treated like an unwrapped tensor

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/functorch/ADInterpreters.cpp:
```diff
@@ -118,6 +119,23 @@ static void autogradBasedTransformSendToNext(
     stack->push_back((*stack)[front + arg_idx]);
   }
+
+  std::vector<int64_t> immutable_inputs;  // all immutable inputs, sorted
+  for (auto idx = stack->size() - args_size; idx < stack->size(); idx++) {
+    const auto ivalue = (*stack)[idx];
+    if (!ivalue.isTensor()) {
+      continue; // only input that can be aliased is a tensor, not a tensor list (expect in ops without returns)
+    }
+    const auto tensor = ivalue.toTensor();
+    auto* maybe_tensor_wrapper = maybeGetTensorWrapper(tensor);
+    if (!maybe_tensor_wrapper || maybe_tensor_wrapper->is_immutable()) {
+      // if the input is immutable, we note its relative position in schema, noting that
+      // args are in reverse order on stack, so the last arg is at the top of the stack
+      const auto relative_pos = idx - (stack->size() - args_size);
+      immutable_inputs.push_back(relative_pos);
+    }
+  }
```
There is some intermediate ground where we keep the existing code in this PR and rely on how std::vector doesn't allocate memory (in most implementations) if we do not populate it. In this world we would only populate `immutable_inputs` if the output has any alias annotations.

samdow replied:
Switched to a bitset--one int64_t is passed around everyone

zou3519 reviewed (commented):
Not done reading through the code yet, will resume soon

Commit:
Update on "fix grad silent correctness issue from view fn followed by an inplace fn"


From https://github.com/pytorch/functorch/issues/1007, which was an issue where we would wrap aliases of unwrapped tensors and miss the inplace error message where we should have gotten it. Instead of keeping aliases unwrapped like I had originally wanted, this simplifies it slightly such that:
(1) All tensors that were previously wrapped are still wrapped. This is occasionally important because of the 1-1 relationship between a tensor and autograd meta. By keeping the same number of wrapper tensors before, we'll never have autograd try to write multiple autograd metas to the same tensor when it wouldn't before
(2) The tensors that either were unwrapped tensors or aliases of unwrapped tensors now get a flag on them (now called `alias_of_unwrapped`). This way, they are still wrapper tensors (and don't have to potentially break autograd) but we can identify that they should be treated like an unwrapped tensor

[ghstack-poisoned]

samdow commented on aten/src/ATen/functorch/ADInterpreters.cpp:
```diff
@@ -31,14 +31,14 @@ static Tensor materializeGradWrappers(const Tensor& tensor, int64_t current_leve
   auto* wrapper = maybeGetTensorWrapper(tensor);
   if (!wrapper) {
-    return makeTensorWrapper(tensor, current_level);
+    return makeTensorWrapper(tensor, current_level, true);
```
Done!

samdow commented on aten/src/ATen/functorch/DynamicLayer.h:
```diff
@@ -115,6 +118,9 @@ TORCH_API std::ostream& operator<<(std::ostream& os, const std::vector<DynamicLa
 
+// used to flag if this is in user code during (not from functorch setup code)
+TORCH_API void setDuringFunctorchTransform(bool during_transform);
+TORCH_API bool getDuringFunctorchTransform();
```
Yep! Good catch

samdow commented on aten/src/ATen/functorch/DynamicLayer.cpp:
```diff
@@ -294,10 +293,24 @@ Tensor unwrapIfDead(const Tensor& tensor) {
+}
+
+void foreachTensorInplaceWithFlag(std::vector<IValue>& args, int64_t begin, int64_t end,
+    std::vector<int64_t> use_flag_relative, std::function<Tensor(const Tensor&, bool)> func){
```
Thanks! Made the new int inputs constant

zou3519 commented on aten/src/ATen/functorch/ADInterpreters.cpp:
```diff
@@ -118,6 +119,24 @@ static void autogradBasedTransformSendToNext(
     stack->push_back((*stack)[front + arg_idx]);
   }
+
+  int64_t immutable_inputs = 0;  // a bitset of the immutable inputs
```
Use `std::bitset<64>` instead of raw int64_t for type-safety; this will also make the code nicer to read (there are APIs like bitset[bit] = ... and bitset::any() that make it so we don't need to do all the bit manipulation below).

zou3519 commented on aten/src/ATen/functorch/ADInterpreters.cpp:
```diff
@@ -136,12 +155,21 @@ static void autogradBasedTransformSendToNext(
 
   // Step 4, 5, 6
   auto ret_size = op.schema().returns().size();
+
+  op.callBoxed(stack);
+
+  // lift_fresh: it's must be freshly allocated and should be wrapped. User shouldn't have access to input version
+  // alias: this is needed for the CompositeImplicit instance norm (running_mean/var get set to be a wrapped value)
+  //        It's not a user facing function, but is more prone to possible errors
+  int64_t outputs_aliasing_immutable = 0;
+  if (!grad_special_case || immutable_inputs == 0) {
+    outputs_aliasing_immutable = findAliasedOutputs(op.schema(), immutable_inputs);
+  }
```
nit: There's no reason to split the `immutable_inputs` computation code and the `outputs_aliasing_immutable` computation code, right? Putting them together would group the logic in a way that we could refactor it into a helper function that, given the inputs and the schema, computes which outputs should be immutable.

zou3519 commented on aten/src/ATen/functorch/Interpreter.h:
```diff
@@ -174,6 +174,17 @@ struct Interpreter {
+//   if use_flag_relative[i] == 1: <-- treats use_flag_relative as a bitset
+//     args[i] = func(args[i], i - begin, true)
+//   args[i] = func(args[i], i - begin)
+// NOTE: relative_skips must be sorted
```
No need to sort a bitset

zou3519 reviewed (approved):
LGTM, left some minor comments inline. TL;DR on those is: please change the int64_t into a `std::bitset<64>`

Commit:
Update on "fix grad silent correctness issue from view fn followed by an inplace fn"


From https://github.com/pytorch/functorch/issues/1007, which was an issue where we would wrap aliases of unwrapped tensors and miss the inplace error message where we should have gotten it. Instead of keeping aliases unwrapped like I had originally wanted, this simplifies it slightly such that:
(1) All tensors that were previously wrapped are still wrapped. This is occasionally important because of the 1-1 relationship between a tensor and autograd meta. By keeping the same number of wrapper tensors before, we'll never have autograd try to write multiple autograd metas to the same tensor when it wouldn't before
(2) The tensors that either were unwrapped tensors or aliases of unwrapped tensors now get a flag on them (now called `alias_of_unwrapped`). This way, they are still wrapper tensors (and don't have to potentially break autograd) but we can identify that they should be treated like an unwrapped tensor

[ghstack-poisoned]

Commit:
Update on "fix grad silent correctness issue from view fn followed by an inplace fn"


From https://github.com/pytorch/functorch/issues/1007, which was an issue where we would wrap aliases of unwrapped tensors and miss the inplace error message where we should have gotten it. Instead of keeping aliases unwrapped like I had originally wanted, this simplifies it slightly such that:
(1) All tensors that were previously wrapped are still wrapped. This is occasionally important because of the 1-1 relationship between a tensor and autograd meta. By keeping the same number of wrapper tensors before, we'll never have autograd try to write multiple autograd metas to the same tensor when it wouldn't before
(2) The tensors that either were unwrapped tensors or aliases of unwrapped tensors now get a flag on them (now called `alias_of_unwrapped`). This way, they are still wrapper tensors (and don't have to potentially break autograd) but we can identify that they should be treated like an unwrapped tensor

[ghstack-poisoned]

Commit:
Update on "fix grad silent correctness issue from view fn followed by an inplace fn"


From https://github.com/pytorch/functorch/issues/1007, which was an issue where we would wrap aliases of unwrapped tensors and miss the inplace error message where we should have gotten it. Instead of keeping aliases unwrapped like I had originally wanted, this simplifies it slightly such that:
(1) All tensors that were previously wrapped are still wrapped. This is occasionally important because of the 1-1 relationship between a tensor and autograd meta. By keeping the same number of wrapper tensors before, we'll never have autograd try to write multiple autograd metas to the same tensor when it wouldn't before
(2) The tensors that either were unwrapped tensors or aliases of unwrapped tensors now get a flag on them (now called `alias_of_unwrapped`). This way, they are still wrapper tensors (and don't have to potentially break autograd) but we can identify that they should be treated like an unwrapped tensor

[ghstack-poisoned]

samdow commented:
@pytorchbot merge

Commit:
Update on "fix grad silent correctness issue from view fn followed by an inplace fn"


From https://github.com/pytorch/functorch/issues/1007, which was an issue where we would wrap aliases of unwrapped tensors and miss the inplace error message where we should have gotten it. Instead of keeping aliases unwrapped like I had originally wanted, this simplifies it slightly such that:
(1) All tensors that were previously wrapped are still wrapped. This is occasionally important because of the 1-1 relationship between a tensor and autograd meta. By keeping the same number of wrapper tensors before, we'll never have autograd try to write multiple autograd metas to the same tensor when it wouldn't before
(2) The tensors that either were unwrapped tensors or aliases of unwrapped tensors now get a flag on them (now called `alias_of_unwrapped`). This way, they are still wrapper tensors (and don't have to potentially break autograd) but we can identify that they should be treated like an unwrapped tensor

[ghstack-poisoned]

samdow commented:
@pytorchbot merge

samdow commented:
@pytorchbot merge -f "failure is OOM on CUDA--unrelated to functorch change and passed on previous commit"
